lib/server.js

```diff
@@ -61,7 +63,11 @@ function createServer (options, httpHandler) {
     if (Object.prototype.hasOwnProperty.call(listenOptions, 'host') === false) {
       listenOptions.host = host
     }
-
+    const serverFactoryListening = server.listening && options.serverFactory
```

Comment:
As we are just looking to throw if the primary server is already `listening`, the `options.serverFactory` check does not seems to be necessary, wdyt?

Reply:
But does this always happen if the first server is listening and `options.serverFactory` is not set?

Reply:
Yeah, it can happen, as [`getServerInstance`](https://github.com/fastify/fastify/pull/4741/files#diff-1b48bb429de959672dadab9c87c215e19177dd70e3bd932d131ce74294c44cd5R25) has been already called, which checks and returns the given server instance.

Then, checking for `options.serverFactory` at this stage becomes slightly redundant

Reply:
Looks like we have to check if `options.serverFactory` exists otherwise it throws an error for Fastify instances without serverFactory as well. This got caught while testing. See the testing logs below
```log
 FAIL ​ test/listen.test.js
 ✖ Server instance returned by 'serverFactory' is already listening

    if (serverFactoryListening) {
      throw new FST_ERR_SERVER_ALREADY_LISTENING()
------------^
    }
    if (host === 'localhost') {

  test: double listen errors
  stack: |
    Object.listen (lib/server.js:69:13)
    cb (test/listen.test.js:138:13)
    onListen (lib/server.js:81:11)
    Object.cb (lib/server.js:142:15)
    Server.wrap (lib/server.js:186:21)
  at:
    line: 69
    column: 13
    file: lib/server.js
    function: Object.listen
  type: FastifyError
  code: FST_ERR_SERVER_ALREADY_LISTENING
  statusCode: 500
  tapCaught: uncaughtException

​ FAIL ​ test/listen.test.js
 ✖ Server instance returned by 'serverFactory' is already listening

    if (serverFactoryListening) {
      throw new FST_ERR_SERVER_ALREADY_LISTENING()
------------^
    }
    if (host === 'localhost') {

  test: double listen errors callback with (err, address)
  stack: |
    Object.listen (lib/server.js:69:13)
    cb (test/listen.test.js:152:13)
    onListen (lib/server.js:81:11)
    Object.cb (lib/server.js:142:15)
    Server.wrap (lib/server.js:186:21)
  at:
    line: 69
    column: 13
    file: lib/server.js
    function: Object.listen
  type: FastifyError
  code: FST_ERR_SERVER_ALREADY_LISTENING
  statusCode: 500
  tapCaught: uncaughtException

​ FAIL ​ test/listen.test.js
 ✖ Server instance returned by 'serverFactory' is already listening

  test/listen.test.js                                                 
  245 |         })                                                    
  246 |     })                                                        
> 247 |     .catch(err => t.error(err))                               
      | --------------------^                                         
  248 | })                                                            
  249 |                                                               
  250 | test('double listen without callback with (address)', t => {  

  test: test/listen.test.js double listen without callback rejects
  stack: |
    test/listen.test.js:247:21

  Error Origin:

    if (serverFactoryListening) {
      throw new FST_ERR_SERVER_ALREADY_LISTENING()
------------^
    }
    if (host === 'localhost') {

  at:
    line: 69
    column: 13
    file: lib/server.js
    function: Object.listen
  stack: |
    Object.listen (lib/server.js:69:13)
    test/listen.test.js:242:15
  type: FastifyError
  code: FST_ERR_SERVER_ALREADY_LISTENING
  statusCode: 500
  message: Server instance returned by 'serverFactory' is already listening

​ FAIL ​ test/listen.test.js
 ✖ Server instance returned by 'serverFactory' is already listening

  test/listen.test.js                                                          
  260 |         })                                                             
  261 |     })                                                                 
> 262 |     .catch(err => t.error(err))                                        
      | --------------------^                                                  
  263 | })                                                                     
  264 |                                                                        
  265 | test('listen twice on the same port without callback rejects', t => {  

  test: test/listen.test.js double listen without callback with (address)
  stack: |
    test/listen.test.js:262:21

  Error Origin:

    if (serverFactoryListening) {
      throw new FST_ERR_SERVER_ALREADY_LISTENING()
------------^
    }
    if (host === 'localhost') {

  at:
    line: 69
    column: 13
    file: lib/server.js
    function: Object.listen
  stack: |
    Object.listen (lib/server.js:69:13)
    test/listen.test.js:257:15
  type: FastifyError
  code: FST_ERR_SERVER_ALREADY_LISTENING
  statusCode: 500
  message: Server instance returned by 'serverFactory' is already listening
​ FAIL ​ test/listen.test.js 4 failed of 51 1s
 ✖ Server instance returned by 'serverFactory' is already listening
 ✖ Server instance returned by 'serverFactory' is already listening
 ✖ Server instance returned by 'serverFactory' is already listening
 ✖ Server instance returned by 'serverFactory' is already listening
```

Reply:
Hmm, not so sure how they are related. But the `getServerInstance` validates if a `options.serverFactory` has been passed, and executes it if so.

If at the first time, we call `getServerInstance` (and internally `opts.serverFactory`, the returned server is already listening. We should throw the error. If a second time we call `getServerInstance` returns a server that is already listening we just emit a warning event.

Maybe worth it to dig deep, but the check should be ok to be done in that way

Reply:
> Maybe worth it to dig deep, but the check should be ok to be done in that way

Looks like the reason it fails those tests is because there are other cases wherein a regular http server gets created. but for some reason it starts listening. I'll try to dig deeper and will raise an issue.